### PR TITLE
Issue #5/skip cache write for some responses

### PR DESCRIPTION
**Pull request by ColinDKelley**
### Types of Changes

Address issue #5 

### Testing

- [ ] I added tests for my changes.

### CODE REVIEW

File: lib/async/http/cache/response.rb [hunk 1]
```diff
@@ -30,33 +30,33 @@ class Response < ::Protocol::HTTP::Response
...
 				attr :generated_at
-				
+
 				def etag
 					@etag ||= @headers[ETAG]
 				end
-				
-				def cachable?
+
+				def cachable? # Spelling should be cacheable?, shouldn't it? -Colin
```
Comment (ColinDKelley) [inline lib/async/http/cache/response.rb]: @ioquatix Ok if I change this spelling?

File: spec/async/http/cache/general_spec.rb [hunk 2]
```diff
@@ -68,43 +68,113 @@
...
 		]}
-		
+
 		it "should cache GET requests" do
 			2.times do
 				user_agents.each do |user_agent|
 					response = subject.get("/", {'user-agent' => user_agent})
 					expect(response.headers['vary']).to include('user-agent')
-					expect(response.read).to be == user_agent
+					expect(response.read).to eq(user_agent)
```
Comment (ColinDKelley) [inline spec/async/http/cache/general_spec.rb]: I've never seen `be ==`. Is that preferred in this project vs simply `eq()`?

File: spec/async/http/cache/general_spec.rb [hunk 3]
```diff
@@ -68,43 +68,113 @@
...
 				end
 			end
-			
-			expect(store.index.size).to be 2
+
+			expect(store.index.size).to eq(2)
```
Comment (ColinDKelley) [inline spec/async/http/cache/general_spec.rb]: I don't think `be 2` was correct, because that's checking for identity.

If not `eq(2)` then `be == 2`.

File: spec/async/http/cache/general_spec.rb [hunk 4]
```diff
@@ -68,43 +68,113 @@
...
 		end
-	end
-	
+  end
+
+  context 'cache writes' do
+    context 'by response code' do
+      let(:app) do
+        Protocol::HTTP::Middleware.for do |_request|
+          Protocol::HTTP::Response[response_code, []] # no response body?
```
Comment (ColinDKelley) [inline spec/async/http/cache/general_spec.rb]: I would like to include a response body because that's typical. But I couldn't get the tests to pass, due to Async waiting and ultimately timing out:
```
Waiting for Async::HTTP::Protocol::HTTP11 pool to drain: #<Async::Pool::Controller(2/∞) 1/1/1;1/1/1>
```
I could use some help with that.

File: spec/async/http/cache/general_spec.rb [hunk 5]
```diff
@@ -68,43 +68,113 @@
...
+        end
+      end
+
+      # [200, 203, 300, 301, 302, 404, 410].each do |response_code|
```
Comment (ColinDKelley) [inline spec/async/http/cache/general_spec.rb]: The above list is accurate per the spec. Current code only caches 200, though. Should I leave it that way or change the code and uncomment the test to match the spec?

File: spec/async/http/cache_spec.rb [hunk 6]
```diff
@@ -1,6 +1,6 @@
 
 RSpec.describe Async::HTTP::Cache do
 	it "has a version number" do
-		expect(Async::HTTP::Cache::VERSION).not_to be nil
+		expect(Async::HTTP::Cache::VERSION).not_to be_empty
```
Comment (ColinDKelley) [inline spec/async/http/cache_spec.rb]: This seemed like a minor improvement so that a VERSION of `""` wouldn't pass.

File: spec/async/http/cache/general_spec.rb (hunk 3 above)
Comment (ioquatix) [reply spec/async/http/cache/general_spec.rb]: Integers are value types so it won't matter, but I'm fine with `== 2`.

File: spec/async/http/cache/general_spec.rb (hunk 2 above)
Comment (ioquatix) [reply spec/async/http/cache/general_spec.rb]: Yeah, I don't like abbreviations and don't find `eq` very intuitive.

File: lib/async/http/cache/response.rb (hunk 1 above)
Comment (ioquatix) [reply lib/async/http/cache/response.rb]: Yeah I'm fine with that.

File: gems.rb [hunk 7]
```diff
@@ -10,4 +10,8 @@
 group :maintenance, optional: true do
 	gem "bake-modernize"
 	gem "bake-bundler"
-end
\ No newline at end of file
+end
+
+group :development do
+  gem "pry"
```
Comment (ioquatix) [inline gems.rb]: I understand the advantage but this is unrelated to the PR :)

File: spec/async/http/cache/general_spec.rb (hunk 5 above)
Comment (ioquatix) [reply spec/async/http/cache/general_spec.rb]: I think it's acceptable to cache more response codes.

File: spec/async/http/cache_spec.rb (hunk 6 above)
Comment (ioquatix) [reply spec/async/http/cache_spec.rb]: Actually, I've been rewriting these to be `.to be =~ /\d+\.\d+\.\d+/` or something like that, just to be more specific.

File: spec/async/http/cache/general_spec.rb (hunk 4 above)
Comment (ioquatix) [reply spec/async/http/cache/general_spec.rb]: I'll take a look.

File: spec/async/http/cache/general_spec.rb (hunk 3 above)
Comment (ColinDKelley) [reply spec/async/http/cache/general_spec.rb]: (Agreed that it will work on that technicality. But it tripped me up to see the explicit identity check because then I wondered why that was important?)

File: spec/async/http/cache/general_spec.rb (hunk 5 above)
Comment (ColinDKelley) [reply spec/async/http/cache/general_spec.rb]: Great, I'll extend it that way.

BTW, this could lead to a philosophical question about whether tests should be DRY with the implementation. I've definitely seen tests (including from new coworkers) that DRY up the tests+implementation by having a constant in the implementation and refer to that from the test. I don't believe in DRYing up code that much because it precludes TDD and the tests will always pass even if the constant gets changed. So instead I'll have a the list of cached response codes duplicated here and in the implementation.

File: spec/async/http/cache_spec.rb (hunk 6 above)
Comment (ColinDKelley) [reply spec/async/http/cache_spec.rb]: Makes sense. I'll do that here.

In specs, do you prefer `=~ /.../` to `match(/.../)`?. I've always used the latter and I think it reads well. (And ironically, in our implementations we've moved away from `=~ /.../` in favor of `.match?(/.../)` based on learning that `=~` was slower because it returns a full `MatchData` object.)

File: gems.rb (hunk 7 above)
Comment (ColinDKelley) [reply gems.rb]: Well, it was related this way:  I needed to use `pry` to figure out some bugs in my specs. :)

Are you saying you want to see a separate PR with this, or that you'd rather not have this development dependency at all?

File: spec/async/http/cache_spec.rb (hunk 6 above)
Comment (ioquatix) [reply spec/async/http/cache_spec.rb]: I don’t have a strong opinion either way.

File: spec/async/http/cache/general_spec.rb (hunk 5 above)
Comment (ioquatix) [reply spec/async/http/cache/general_spec.rb]: I’m fine with either approach.

File: gems.rb (hunk 7 above)
Comment (ioquatix) [reply gems.rb]: I think adding pry as a development dependency is outside of the scope of this PR.

I don’t have a strong opinion about introducing pry. I usually just use `binding.irb` these days. I used to use pry a lot but don’t tend to use it much these days.

**Comment by ioquatix** [comment]
I've fixed the test suite do you mind rebasing? Thanks!

**Comment by ioquatix** [comment]
I'm still seeing some issues on Linux, I only have my laptop in bed, I'll need to check it on my Linux desktop tomorrow.

File: lib/async/http/cache/response.rb (hunk 1 above)
Comment (ioquatix) [reply lib/async/http/cache/response.rb]: I fixed it so you can just rebase.

File: spec/async/http/cache/general_spec.rb (hunk 4 above)
Comment (ioquatix) [reply spec/async/http/cache/general_spec.rb]: Can you please rebase and try again?

File: spec/async/http/cache/general_spec.rb [hunk 8]
```diff
@@ -70,43 +70,112 @@
...
 		end
-	end
-	
+  end
+
+  context 'cache writes' do
+    context 'by response code' do
+      let(:app) do
+        Protocol::HTTP::Middleware.for do |_request|
+          Protocol::HTTP::Response[response_code, [], ['body']]
```
Comment (ColinDKelley) [inline spec/async/http/cache/general_spec.rb]: @ioquatix This is the test where you suggested I rebase and try again. I have now, but I'm still getting a timeout here when `body` isn't empty:
```
$ bundle exec rspec spec/async/http/cache/general_spec.rb
...
    cache writes
      by response code
        when cacheable: 200
 0.19s     warn: Async::HTTP::Client [oid=0x1bbc] [ec=0x1bd0] [pid=14740] [2022-08-22 19:32:36 -0700]
               | Waiting for Async::HTTP::Protocol::HTTP1 pool to drain: #<Async::Pool::Controller(2/∞) 1/1/1;1/1/1>
          is cached (FAILED - 1)
        when cacheable: 203
10.18s     warn: Async::HTTP::Client [oid=0x1d74] [ec=0x1d88] [pid=14740] [2022-08-22 19:32:46 -0700]
               | Waiting for Async::HTTP::Protocol::HTTP1 pool to drain: #<Async::Pool::Controller(2/∞) 1/1/1;1/1/1>
          is cached (FAILED - 2)
...
```

Comment (ioquatix) [reply spec/async/http/cache/general_spec.rb]: I'll check what's going on.
